cas-mfa-java/src/main/java/net/unicon/cas/mfa/authentication/principal/ServiceSpecificPrincipalAttributeMultiFactorAuthenticationRequestResolver.java

```diff
@@ -0,0 +1,258 @@
+package net.unicon.cas.mfa.authentication.principal;
+
+import net.unicon.cas.addons.serviceregistry.RegisteredServiceWithAttributes;
+import net.unicon.cas.mfa.authentication.AuthenticationMethodConfigurationProvider;
+import net.unicon.cas.mfa.authentication.MultiFactorAuthenticationRequestContext;
+import net.unicon.cas.mfa.web.support.MultiFactorAuthenticationSupportingWebApplicationService;
+import net.unicon.cas.mfa.web.support.MultiFactorWebApplicationServiceFactory;
+import org.jasig.cas.authentication.Authentication;
+import org.jasig.cas.authentication.principal.WebApplicationService;
+import org.jasig.cas.services.RegisteredService;
+import org.jasig.cas.services.ServicesManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.validation.constraints.NotNull;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import static net.unicon.cas.mfa.web.support.MultiFactorAuthenticationSupportingWebApplicationService.AuthenticationMethodSource;
+
+/**
+ * Implementation of <code>MultiFactorAuthenticationRequestResolver</code> that resolves
+ * potential mfa request based on the configured principal attribute.
+ * Note: It is assumed that the attribute value that specifies the
+ * authentication method at this time is a single-valued attribute.
```

Comment:
No this came from the parent class I started with... I'll clean up the JavaDoc on the new version.